### PR TITLE
consider TracerSpan assoc type -> Span

### PR DESCRIPTION
**Pull request by ktoso**
Something @fabianfett requested we give another look at.

Resolves https://github.com/apple/swift-distributed-tracing/issues/110 but we should discuss this more -- it comes with a confusing tradeoff for implementers of spans I suppose. Please discuss. I'm on the fence, it is nice to not have `Tracer.TracerSpan` but the tradeoff is some potential confusion hmmm

**Motivation:**

Users who use generic `Tracer` end up having to spell `Tracer.TracerSpan`.

We could help those situations by renaming the associated type. However at the cost of causing name conflicts where the Span implementation must qualify that it is implementing `Tracing.Span`.

This may or may not be worth it -- discuss.

**Modifications:**

`associatedtype TracerSpan: Span` -> `associatedtype Span: Tracing.Span`

### CODE REVIEW

**Comment by fabianfett** [comment]
@ktoso can you give an example for this?

> We could help those situations by renaming the associated type. However at the cost of causing name conflicts where the Span implementation must qualify that it is implementing Tracing.Span.

File: Sources/Tracing/Tracer.swift [hunk 1]
```diff
@@ -59,7 +59,7 @@ extension Tracer {
         function: String = #function,
         file fileID: String = #fileID,
         line: UInt = #line
-    ) -> any Span {
```
[the remark this answers is not in the document]
Comment (ktoso) [reply Sources/Tracing/Tracer.swift:62]: Yeah here I can likely just do Span, that's right.

In a `TracerProtocol` we sometimes had to be explicit.

I can give this a cleanup pass

File: Sources/Tracing/NoOpTracer.swift [hunk 2]
```diff
@@ -48,7 +48,7 @@ public struct NoOpTracer: LegacyTracerProtocol {
         // no-op
     }
 
-    public struct NoOpSpan: Span {
+    public struct NoOpSpan: Tracing.Span {
```
Comment (ktoso) [inline Sources/Tracing/NoOpTracer.swift:51]: This is one annoying spot, people MUST write Tracing.Span here

Comment (stevapple) [reply Sources/Tracing/NoOpTracer.swift:51]: What about swapping `TracerSpan` and `Span`?

Comment (ktoso) [reply Sources/Tracing/NoOpTracer.swift:51]: Maybe, what do @tomerd @fabianfett @Lukasa think?

Comment (fabianfett) [reply Sources/Tracing/NoOpTracer.swift:51]: I don't find this to annoying. The tough part here is really only on the tracer implementers. TracerSpan is on all the users. If we want to keep the pain small, the first group should deal with it, not the second.

Comment (ktoso) [reply Sources/Tracing/NoOpTracer.swift:51]: Yeah it's on implementors to do the dance, I don't mind either I guess.

Comment (stevapple) [reply Sources/Tracing/NoOpTracer.swift:51]: I would prefer `TracerSpan` to `[Tracing.]Span` here. Swift’s name lookup is still tricky now (see https://forums.swift.org/t/pitch-fully-qualified-name-syntax/28482 for the context). If the client defines a top-level `Tracing` type, they will completely lose access to `Tracing.Span`. Not to say `TracerSpan` is shorter by two ASCII characters:)

**Comment by ktoso** [comment]
Yeah I think this will be fine. I'll merge tomorrow probably tho

File: Sources/Tracing/NoOpTracer.swift (hunk 2 above)
Comment (stevapple) [reply Sources/Tracing/NoOpTracer.swift:51]: A typical tracing user doesn’t use either `[Tracing.]Span` or `Tracer.Span`, as the `Tracer.Span` type can be inferred from `Tracer.startSpan`.

Both `Span` and `TracerSpan` can cause some confusion with `Tracer.Span`, but the fact that `Span` is ambiguous, despite being possible to work around and comfort the compiler, will inevitably troubles code review.

Given that the type name is not used very often, I still suggests the longer but less error-prone `TracerSpan`.

**Comment by ktoso** [comment]
Let's give this a shot

**Pull request merged**